pkg/controller/mysqlcluster/internal/cleaner/pvc.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2018 Platform9 Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysqlcluster
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	api "github.com/presslabs/mysql-operator/pkg/apis/mysql/v1alpha1"
+	wrapcluster "github.com/presslabs/mysql-operator/pkg/controller/internal/mysqlcluster"
+	"github.com/presslabs/mysql-operator/pkg/options"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+const (
+	controllerName       = "controller.mysqlcluster"
+	deleteSuccess        = "SucessfulDelete"
+	deleteFail           = "FailedDelete"
+	messagePvcDeleted    = "delete Claim %s in StatefulSet %s successful"
+	messagePvcNotDeleted = "delete Claim %s in StatefulSet %s failed"
+)
+
+var log = logf.Log.WithName(controllerName)
+
+// PvcCleaner represents an object to clean Pvcs of a MysqlCluster
+type PvcCleaner struct {
+	cluster *wrapcluster.MysqlCluster
+	opt     *options.Options
+	sts     *apps.StatefulSet
+}
+
+// NewPvcCleaner returns a new PVC cleaner object
+func NewPvcCleaner(cluster *api.MysqlCluster, opt *options.Options) *PvcCleaner {
+	obj := &apps.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.GetNameForResource(api.StatefulSet),
+			Namespace: cluster.Namespace,
+		},
+	}
+	return &PvcCleaner{
+		cluster: wrapcluster.NewMysqlClusterWrapper(cluster),
+		opt:     opt,
+		sts:     obj,
+	}
+}
+
+// Run performs cleanup of orphaned pvcs in a Mysql cluster
+func (p *PvcCleaner) Run(ctx context.Context, c client.Client, scheme *runtime.Scheme, recorder record.EventRecorder) error {
+	sts := p.sts
+	stsMeta := sts.ObjectMeta
+
+	key := types.NamespacedName{Name: stsMeta.GetName(), Namespace: stsMeta.GetNamespace()}
+
+	err := c.Get(ctx, key, sts)
+	if err != nil {
+		return err
+	}
+
+	if sts.DeletionTimestamp != nil {
+		log.V(4).Info("being deleted, no action", "Statefulset", key)
+		return nil
+	}
+
+	if len(sts.Spec.VolumeClaimTemplates) == 0 {
+		log.V(4).Info("ignored, no volume claims", "Statefulset", key)
+		return nil
+	}
+
+	replicas := *sts.Spec.Replicas
```

Comment:
The number of desired replicas should be checked against the `MysqlCluster` rather than the `StatefulSet`, as this is the source of truth. Other than this, the PR looks good.